pkg/operator/controller/controller_dns.go

```diff
@@ -1,89 +1,139 @@
 package controller
 
 import (
+	"context"
 	"fmt"
 
+	iov1 "github.com/openshift/cluster-ingress-operator/pkg/api/v1"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
 	operatorv1 "github.com/openshift/api/operator/v1"
-	"github.com/openshift/cluster-ingress-operator/pkg/dns"
 
 	corev1 "k8s.io/api/core/v1"
 
-	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // ensureDNS will create DNS records for the given LB service. If service is
 // nil, nothing is done.
-func (r *reconciler) ensureDNS(ci *operatorv1.IngressController, service *corev1.Service, dnsConfig *configv1.DNS) error {
-	records := desiredDNSRecords(ci, dnsConfig, service)
-	for _, record := range records {
-		err := r.DNSManager.Ensure(record)
-		if err != nil {
-			return fmt.Errorf("failed to ensure DNS record %v for %s/%s: %v", record, ci.Namespace, ci.Name, err)
-		}
-		log.Info("ensured DNS record for ingresscontroller", "namespace", ci.Namespace, "name", ci.Name, "record", record)
+func (r *reconciler) ensureWildcardDNSRecord(ic *operatorv1.IngressController, service *corev1.Service) (*iov1.DNSRecord, error) {
+	if service == nil {
+		return nil, nil
 	}
-	return nil
-}
 
-func newAliasRecord(domain, target string, zone configv1.DNSZone) *dns.Record {
-	return &dns.Record{
-		Zone: zone,
-		Type: dns.ALIASRecord,
-		Alias: &dns.AliasRecord{
-			Domain: domain,
-			Target: target,
-		},
+	desired := desiredWildcardRecord(ic, service)
+	current, err := r.currentWilcardDNSRecord(ic)
+	if err != nil {
+		return nil, err
 	}
-}
-
-func newARecord(domain, target string, zone configv1.DNSZone) *dns.Record {
-	return &dns.Record{
-		Zone: zone,
-		Type: dns.ARecordType,
-		ARecord: &dns.ARecord{
-			Domain:  domain,
-			Address: target,
-		},
+	if desired != nil && current == nil {
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return nil, fmt.Errorf("failed to create dnsrecord %s/%s: %v", desired.Namespace, desired.Name, err)
+		}
+		log.Info("created dnsrecord", "namespace", desired.Namespace, "dnsrecord", desired)
+		return desired, nil
 	}
+	return current, nil
 }
 
-// desiredDNSRecords will return any necessary DNS records for the given inputs.
-// If an ingress domain is in use, records are desired in every specified zone
-// present in the cluster DNS configuration.
-func desiredDNSRecords(ci *operatorv1.IngressController, dnsConfig *configv1.DNS, service *corev1.Service) []*dns.Record {
-	records := []*dns.Record{}
-
+// desiredDNSRecords will return any necessary wildcard DNS records for the
+// ingresscontroller.
+//
+// For now, if the service has more than one .status.loadbalancer.ingress, only
+// the first will be used.
+//
+// TODO: If .status.loadbalancer.ingress is processed once as non-empty and then
+// later becomes empty, what should we do? Currently we'll treat it as an intent
+// to not have a desired record.
```

Comment:
`desiredWildcardRecord` will return nil, but `ensureWildcardDNSRecord` will not delete the current record if `desired` is nil, which is what I was getting at when I suggested adding a comment.

Reply:
delete is deliberately not handled during ensure in this case, so that part... I think mutations of .status.ingress ingress are a larger topic at this point. Unless we compute desired from a persisted "last observed" state I think there are probably more than one of these edge cases we don't coherently account for...